libs/component-tests/src/main/java/org/codice/ddf/test/component/UrlBuilder.java

```diff
@@ -0,0 +1,102 @@
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.test.component;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UrlBuilder {
+    private static final Logger LOGGER = LoggerFactory.getLogger(UrlBuilder.class);
+
+    private final URL rootContextUrl;
+
+    private final URL secureRootContextUrl;
+
+    public static UrlBuilder from(String hostname, String port, String securePort,
+            String rootContext) {
+        return new UrlBuilder(hostname, port, securePort, rootContext);
+    }
+
+    public static UrlBuilder fromSystemProperties(String hostnameSystemProperty,
+            String portSystemProperty, String securePortSystemProperty,
+            String rootContextSystemProperty) {
+
+        return new UrlBuilder(System.getProperty(hostnameSystemProperty),
+                System.getProperty(portSystemProperty),
+                System.getProperty(securePortSystemProperty),
+                System.getProperty(rootContextSystemProperty));
+    }
+
+    public UrlBuilder add(String contextPath) {
+        String cleanContextPath = cleanContextPath(contextPath);
+
+        try {
+            return new UrlBuilder(//
+                    new URL(String.format("%s/%s", rootContextUrl.toString(), cleanContextPath)),
+                    new URL(String.format("%s/%s",
+                            secureRootContextUrl.toString(),
+                            cleanContextPath)));
+        } catch (MalformedURLException e) {
+            String message = "Cannot build valid URL using context path " + contextPath;
+            LOGGER.error(message, e);
+            throw new IllegalArgumentException(message, e);
+        }
+    }
+
+    public URL build() {
+        return rootContextUrl;
+    }
+
+    public URL buildSecure() {
+        return secureRootContextUrl;
+    }
+
+    private UrlBuilder(URL rootContextUrl, URL secureRootContextUrl) {
+        this.rootContextUrl = rootContextUrl;
+        this.secureRootContextUrl = secureRootContextUrl;
+        LOGGER.debug("UrlBuilder created for {}, {}", rootContextUrl, secureRootContextUrl);
+    }
+
+    private UrlBuilder(String hostname, String port, String securePort, String rootContext) {
+
+        try {
+            this.rootContextUrl = new URL(String.format("http://%s:%s/%s",
```

Comment:
❓ Some of the refactoring I mentioned above could also clean up this code down here. We shouldn't need to do the same construction, validation, and logging in multiple places.